docs/rfc/007-endpoint-errors.md

```diff
@@ -0,0 +1,229 @@
+    .blueMoon(e -> handleBlueMoonEvent())
+    .build());
+    
+// v- alternative implementation where you force logic splitting
+myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), MyConjureServiceErrors.getSomeResultBuilder()
+    .blueMoon(e -> handleBlueMoonEvent())
+    .ok(mySafeResult3 -> handleOk(mySafeResult3))
+    .build());
+    
+// v- or via `Visitor<T>` and return `T` in all branches
+SomeOtherResult otherResult = myConjureService.getSomeResult(AuthHeader.valueOf("authHeader"), MyConjureServiceErrors.getSomeResultBuilder()
+    .blueMoon(e -> handleBlueMoonEvent())
+    .ok(mySafeResult3 -> handleOk(mySafeResult3))
+    .build());
+```
+
+## Proposal
+
+This proposal should be viewed in two distinct parts - declared (and handled) exceptions on Conjure endpoints, and the 
+language implementations of such a feature.
+
+### Intent
+
+The primary intent of this proposal is to provide more expressivity around errors in a Conjure API, while minimizing the 
+change surface-area, especially around wire format changes. Our proposed change introduces *no* wire changes, so should 
+be purely additive and not dramatically affect back-compat.
+
+### Declared Errors
+
+**Semantics:** To declare an error on an endpoint is to say “This is a very real outcome of calling this endpoint. You 
+should handle this outcome to ensure that you can react accordingly”. It should be something closely tied to the 
+functionality of that endpoint - that is to say, `getFoo` declaring `FooNotReadyYet` would be a good example, but 
+`UserNotAuthorized` would be undeclared / passed through as normal (as it is a standard expectation of interacting with 
+an authorized service).
+
+**Pros:**
+
+* Consumers now know the *expected* outcomes of calling an endpoint in a first-class way and can handle those outcomes accordingly.
+* Producers are able to express the different failure states of their endpoints in greater detail, and have conviction that those failure states are being handled by consumers.
+* Changing the failure states of an endpoint now visibly affects the API definition - changes in behavior are properly captured in code (rather than just documentation).
+
+**Cons:**
+
+* To be effective, these failure cases will need to be ‘forced’ upon consumers via the compiler. This could be as complex 
+  data types (perhaps unions of {success, error1, error2, ...} or via checked exceptions. This would entail some degree 
+  of (justified) work migrating to a with-errors world to wire each failure case into the correct behavior in client code.
+* Given the new flexibility, producers may more eagerly throw errors in their code, and ‘old’ clients will have a hard 
+  time dealing with these without upgrading.
+* Producers might be incentivized to cram *all* the possible outcomes of their endpoint as errors (see the 
```

Comment:
yeah - how do we capture that the intent is more result-type-esque (imo) vs listing all of the truly 'exceptional' cases